test/Prototypes/string_prototype/StringComparison.swift

```diff
@@ -633,3 +634,393 @@ where
+    return buffer[i]
+  }
+  public func index(after i: Index) -> Index {
+    return buffer.index(after: i)
+  }
+  public func index(before i: Index) -> Index {
+    return buffer.index(before: i)
+  }
+  public typealias SubSequence = BidirectionalSlice<FCCNormalizedSegment>
+}
+
+// Initialize an UnboundCapacity from any sequence
+extension UnboundCapacity {
+  public init<S : Sequence>(_ elements: S)
+  where S.Iterator.Element == Base.Iterator.Element {
+    self.init()
+    self.reserveCapacity(elements.underestimatedCount)
+    for elt in elements {
+      self.append(elt)
+    }
+  }
+}
+
+// Ask ICU if the given unicode scalar value has a normalization boundary before
+// it, that is it begins a new normalization segment.
+public func _hasBoundary(before value: UInt32) -> Bool {
+  return __swift_stdlib_unorm2_hasBoundaryBefore(_fccNormalizer, value) != 0
+}
+
+struct FCCNormalizedLazySegments<
+  CodeUnits : RandomAccessCollection,
+  FromEncoding : UnicodeEncoding
+> : BidirectionalCollection
+where
+  CodeUnits.Index == CodeUnits.SubSequence.Index,
+  CodeUnits.SubSequence : RandomAccessCollection,
+  CodeUnits.SubSequence == CodeUnits.SubSequence.SubSequence,
+  CodeUnits.Iterator.Element == CodeUnits.SubSequence.Iterator.Element,
+  CodeUnits.Iterator.Element == FromEncoding.EncodedScalar.Iterator.Element
+{
+  let codeUnits: CodeUnits
+
+  public init(
+    _ codeUnits: CodeUnits,
+    _: FromEncoding.Type = FromEncoding.self
+  ) {
+    self.codeUnits = codeUnits
+  }
+
+  public init(_ unicodeView: _UnicodeViews<CodeUnits, FromEncoding>) {
+    self.init(unicodeView.codeUnits)
+  }
+
+  private func priorSegment(
+    endingAt end: CodeUnits.Index
+  ) -> Index {
+    _debug("priorSegment endingAt: end")
+    precondition(end != codeUnits.startIndex)
+
+    // Decode scalars backwards, including them until we after we find one that
+    // has a boundary before it (and include that one).
+    var start = end
+    while start != codeUnits.startIndex {
+      // Include this scalar
+      let (scalarValue: value, startIndex: scalarStart, e)
+        = decodeOne(priorTo: start)
+      assert(e == start, "Internal inconsistency in decodeOne")
+
+      // Include this scalar
+      start = scalarStart
+
+      if _hasBoundary(before: value) {
+        // We're done
+        break
+      }
+    }
+
+    return Index(
+      nativeStartIndex: start,
+      nativeEndIndex: end,
+      segment: formSegment(from: start, until: end)
+    )
+  }
+
+  // Scan ahead and produce the next segment. `from` must be valid index (not
+  // end).
+  private func nextSegment(
+    startingAt start: CodeUnits.Index
+  ) -> Index {
+    if start == codeUnits.endIndex {
+      return endIndex
+    }
+
+    // Parse the first scalar, it will always be in the segment
+    var (scalarValue: value, startIndex: s, endIndex: end)
+      = decodeOne(from: start)
+    assert(_hasBoundary(before: value), "Not on segment boundary")
```

Comment:
Is this assertion correct?  What if you're at the beginning of a string that starts with a non-starter?